File: Other/Review/Phase 7/Review4/Part2/monotonic-stack/monotonic-stack.ts
export function nextGreaterElement(nums: number[]): number[] {
  const result: number[] = new Array(nums.length).fill(-1);
  const monotonicStack: number[] = [];

  for (let i = 0; i < nums.length; i++) {
    while (monotonicStack.length && nums[monotonicStack[monotonicStack.length - 1]] < nums[i]) {
      const top = monotonicStack.pop()!;
      result[top] = nums[i];
    }
    monotonicStack.push(i);
  }

  return result;
}

// test
const nums = [2, 1, 2, 4, 3];
console.log(nextGreaterElement(nums)); // [4, 2, 4, -1, -1]

File: Data Structure & Algorithm/Review/Stage 7/Review 4/Part2/monotonic-stack/monotonic-stack.go
package main

import "fmt"

func nextGreaterElement(nums []int) []int {
	monotonicStack := make([]int, 0)
	results := make([]int, len(nums))

	for i := 0; i < len(nums); i++ {
		results[i] = -1
	}

	for i := 0; i < len(nums); i++ {
		for len(monotonicStack) > 0 && nums[monotonicStack[len(monotonicStack)-1]] < nums[i] {
			top := monotonicStack[len(monotonicStack)-1]
			monotonicStack = monotonicStack[:len(monotonicStack)-1]

			results[top] = nums[i]
		}
		monotonicStack = append(monotonicStack, i)
	}

	return results
}

// test
func main() {
	nums := []int{2, 1, 2, 4, 3}
	fmt.Println(nextGreaterElement(nums)) // [4, 2, 4, -1, -1]
}
